### PR TITLE
Benchmark tests in Travis

### LINKED ISSUE #699: Integrate js-framework-benchmark into tests and CI
**Enhancement** 

Based off of discussion in dojo/examples#240, we should consider integrating [`js-framework-benchmark`](https://github.com/krausest/js-framework-benchmark) into the testing of widget-core.

### PR DESCRIPTION
**Pull request by umaar**
**Type:**  feature

The following has been addressed in the PR:

* [x] There is a related issue
* [ ] All code matches the [style guide](https://github.com/dojo/meta/blob/master/STYLE.md)
* [ ] Unit or Functional tests are included in the PR

**Description:**

This PR embeds a modified version of `https://github.com/krausest/js-framework-benchmark` directly into this repository, along with the Dojo 2 sample application which the benchmark tool hooks into. The travis build is instructed to run `npm run benchmark` which in turn triggers a bash script to run the benchmark tests and then finally console.log the results.

Architecture discussions were done with @agubler and happy to implement any feedback/changes people have.

To actually save previous benchmark results, we could create a web service, or use `https://jsonbin.org/` for example. Also, the work to fail the build when a regression occurs will be done at a later stage as agreed with @agubler - this PR only logs the benchmarking results to the console.

Example output: 

```
01_run1k - vanilla: 255.465. dojo: 425.121 (66% slower)
02_replace1k - vanilla: 58.789. dojo: 457.797 (679% slower)
03_update10th1k - vanilla: 103.432. dojo: 238.904 (131% slower)
04_select1k - vanilla: 5.383. dojo: 16.947 (215% slower)
05_swap1k - vanilla: 10.289. dojo: 32.962 (220% slower)
06_remove-one-1k - vanilla: 57.628. dojo: 97.031 (68% slower)
07_create10k - vanilla: 2679.103. dojo: 4098.967 (53% slower)
08_create1k-after10k - vanilla: 359.038. dojo: 621.654 (73% slower)
09_clear10k - vanilla: 321.489. dojo: 448.699 (40% slower)
21_ready-memory - vanilla: 2.68634033203125. dojo: 3.7585678100585938 (40% slower)
22_run-memory - vanilla: 3.2527084350585938. dojo: 9.984672546386719 (207% slower)
23_update5-memory - vanilla: 3.2912979125976562. dojo: 10.110137939453125 (207% slower)
24_run5-memory - vanilla: 3.4715347290039062. dojo: 10.643966674804688 (207% slower)
25_run-clear-memory - vanilla: 3.0312652587890625. dojo: 4.471351623535156 (48% slower)
30_startup - vanilla: 161. dojo: 704 (337% slower)
```

This PR will increase the build time by ~4 minutes.

Resolves #699

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -38,21 +39,30 @@
...
+    "jstat": "^1.7.1",
+    "ramda": "^0.25.0",
+    "selenium-webdriver": "3.6.0",
+    "serve": "^6.4.1",
     "sinon": "^1.17.6",
     "tslint": "^5.2.0",
     "typescript": "~2.5.2"
   },
   "dependencies": {
+    "@types/ramda": "^0.25.5",
```
Comment (agubler) [inline package.json]: Should be a dev dep

File: tests/benchmark/app/App.ts [hunk 2]
```diff
@@ -0,0 +1,124 @@
+
```
Comment (agubler) [inline tests/benchmark/app/App.ts]: Empty line

File: tests/benchmark/app/App.ts [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+import { v, w } from '../../../src/d';
+
+import { Row } from './Row';
+import { Buttons, ButtonConfig } from './Buttons';
+import { Store } from  './Store';
+
+let startTime: number;
+let lastMeasure: string | null;
+
+function startMeasure(name: string) {
```
Comment (agubler) [inline tests/benchmark/app/App.ts]: For our internal impl we don’t need to console log the approx run times

File: tests/benchmark/app/App.ts [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+		{ id: 'clear', label: 'Clear', onClick: this._clear },
+		{ id: 'swaprows', label: 'Swap Rows', onClick: this._swapRows }
+	];
+
+	protected render (): DNode {
+		const { _select, _delete, _store } = this;
+		const rows = _store.data.map(({ id, label }, index) => {
+			return w(Row, {
+				id,
+				key: index,
```
Comment (agubler) [inline tests/benchmark/app/App.ts]: Let’s make this `id` so it’s a true keyed impl as per the benchmarks definition

File: tests/benchmark/app/Button.ts [hunk 5]
```diff
@@ -0,0 +1,24 @@
+
```
Comment (agubler) [inline tests/benchmark/app/Button.ts]: Empty line

File: tests/benchmark/app/index.html [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+	<script src="../../../../node_modules/@dojo/loader/loader.js"></script>
+	<script>
+		require.config({
+			packages: [
+				{ name: '@dojo', location: '../../../../node_modules/@dojo' },
+				{ name: 'pepjs', location: '../../../../node_modules/pepjs/dist', main: 'pep' }
+			]
+		});
+
+		require([ '../../../../_build/tests/benchmark/app/main' ], function () {});
```
Comment (agubler) [inline tests/benchmark/app/index.html]: Is this path right? We should copy it over to _build as part of the grunt task?

File: tests/benchmark/app/css/useMinimalCss.css [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+}
+.table td.col-md-6 {
+    width: 400px;
+}
+.glyphicon-remove:after {
+    content: 'x'
+}
+tr.danger td {
+    background-color: #faa;
+}
```
Comment (agubler) [inline tests/benchmark/app/css/useMinimalCss.css]: Needs new line

File: tests/benchmark/app/css/main.css [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+    position: absolute;
+    top:-20px;
+    left:-20px;
+}
+.col-sm-6.smallpad {
+    padding: 5px;
+}
+.jumbotron .row h1 {
+    font-size: 40px;
+}
```
Comment (agubler) [inline tests/benchmark/app/css/main.css]: Needs new line

File: tests/benchmark/runner/package.json [hunk 9]
```diff
@@ -0,0 +1,38 @@
+{
```
Comment (agubler) [inline tests/benchmark/runner/package.json]: Do we need the seperate package.json?

File: tests/benchmark/runner/tsconfig.json [hunk 10]
```diff
@@ -0,0 +1,26 @@
+{
```
Comment (agubler) [inline tests/benchmark/runner/tsconfig.json]: Do we need another TS config?

Comment (umaar) [reply tests/benchmark/runner/tsconfig.json]: Removed!

File: tests/benchmark/runner/package.json (hunk 9 above)
Comment (umaar) [reply tests/benchmark/runner/package.json]: True, removed!

File: tests/benchmark/app/index.html (hunk 6 above)
Comment (umaar) [reply tests/benchmark/app/index.html]: Fixed, is now using the built version

File: tests/benchmark/app/css/useMinimalCss.css (hunk 7 above)
Comment (umaar) [reply tests/benchmark/app/css/useMinimalCss.css]: Done

File: tests/benchmark/app/css/main.css (hunk 8 above)
Comment (umaar) [reply tests/benchmark/app/css/main.css]: Done

File: tests/benchmark/app/Button.ts (hunk 5 above)
Comment (umaar) [reply tests/benchmark/app/Button.ts]: Done

File: tests/benchmark/app/App.ts (hunk 4 above)
Comment (umaar) [reply tests/benchmark/app/App.ts]: Done

File: tests/benchmark/app/App.ts (hunk 3 above)
Comment (umaar) [reply tests/benchmark/app/App.ts]: Removed all code related to console logging

File: tests/benchmark/app/App.ts (hunk 2 above)
Comment (umaar) [reply tests/benchmark/app/App.ts]: Fixed

File: package.json (hunk 1 above)
Comment (umaar) [reply package.json]: Fixed

**Comment by umaar** [comment]
@agubler ready for another round of review! As we spoke about I made it so the vanilla benchmarks are run as a baseline 👍

File: tests/benchmark/css/github-markdown.css [hunk 11]
```diff
@@ -0,0 +1,2 @@
...ebkit-text-size-adjust:100%;text-size-adjust:100%;color:#333;font-family:"Helvetica Neue",Helvetica,"Segoe UI",Arial,freesans,sans-serif,"Apple Color Emoji","Segoe UI Emoji","Segoe UI Symbol";font-size:16px;line-height:1.6;word-wrap:break-word;}.markdown-body a{background-color:transparent;}.markdown-body a:active,.markdown-body a:hover{outline:0;}.markdown-body strong{font-weight:bold;}.markdown-body h1{font-size:2em;margin:0.67em 0;}.markdown-body img{border:0;}.markdown-body hr{box-sizing:content-box;height:0;}.markdown-body pre{overflow:auto;}.markdown-body code,.markdown-body kbd,.markdown-body pre{font-family:monospace,monospace;font-size:1em;}.markdown-body input{color:inherit;font:inherit;margin:0;}.markdown-body html input[disabled]{cursor:default;}.markdown-body input{line-height:normal;}.markdown-body input[type="checkbox"]{box-sizing:border-box;padding:0;}.markdown-body table{border-collapse:collapse;border-spacing:0;}.markdown-body td,.markdown-body th{padding:0;}.markdown-body *{box-sizing:border-box;}.markdown-body input{font:13px/ 1.4 Helvetica, arial, nimbussansl, liberationsans, freesans, clean, sans-serif, "Apple Color Emoji", "Segoe UI Emoji", "Segoe UI Symbol";
```
Comment (agubler) [inline tests/benchmark/css/github-markdown.css]: Is this used? if so can we use an npm dependency?

File: tests/benchmark/runner/src/benchmarkRunner.ts [hunk 12]
```diff
@@ -0,0 +1,448 @@
...
+				filteredEvents.push({type: 'click', ts: +e.params.ts, dur: +e.params.dur, end: +e.params.ts + e.params.dur});
+			}
+		} else if (e.params.name === 'TimeStamp' &&
+			(e.params.args.data.message === 'afterBenchmark' || e.params.args.data.message === 'finishedBenchmark' || e.params.args.data.message === 'runBenchmark' || e.params.args.data.message === 'initBenchmark')) {
+			filteredEvents.push({type: e.params.args.data.message, ts: +e.params.ts, dur: 0, end: +e.params.ts});
+		} else if (e.params.name === 'navigationStart') {
+			filteredEvents.push({type: 'navigationStart', ts: +e.params.ts, dur: 0, end: +e.params.ts});
+		} else if (e.params.name === 'Paint') {
+			filteredEvents.push({type: 'paint', ts: +e.params.ts, dur: +e.params.dur, end: +e.params.ts + e.params.dur, evt: JSON.stringify(e)});
+		// } else if (e.params.name==='Rasterize') {
```
Comment (agubler) [inline tests/benchmark/runner/src/benchmarkRunner.ts]: let's remove these as they are commented out

File: tests/benchmark/runner/src/benchmarkRunner.ts [hunk 13]
```diff
@@ -0,0 +1,448 @@
...
+			}
+
+			let lastPaint = R.reduce((max: any, elem: any) => max.end > elem.end ? max : elem, {end: 0}, paints);
+
+			const lastEvent: any = R.last(eventsDuringBenchmark);
+			let eventsDuringBenchmarkEnd: any;
+
+			if (lastEvent && lastEvent.end) {
+				eventsDuringBenchmarkEnd = lastEvent.end;
+			} else {
```
Comment (agubler) [inline tests/benchmark/runner/src/benchmarkRunner.ts]: else should be on the next line

File: tests/benchmark/runner/src/benchmarkRunner.ts [hunk 14]
```diff
@@ -0,0 +1,448 @@
...
+	return durationJS;
+}
+
+function buildDriver() {
+	let logPref = new logging.Preferences();
+	logPref.setLevel(logging.Type.PERFORMANCE, logging.Level.ALL);
+	logPref.setLevel(logging.Type.BROWSER, logging.Level.ALL);
+
+	let options = new chrome.Options();
+	// options = options.setChromeBinaryPath("/Applications/Google\ Chrome.app/Contents/MacOS/Google\ Chrome");
```
Comment (agubler) [inline tests/benchmark/runner/src/benchmarkRunner.ts]: can we remove the commented code?

File: tests/benchmark/runner/src/benchmarkRunner.ts [hunk 15]
```diff
@@ -0,0 +1,448 @@
...
+	// options = options.setChromeBinaryPath("/Applications/Chromium.app/Contents/MacOS/Chromium");
+	// options = options.setChromeBinaryPath("/Applications/Google Chrome Canary.app/Contents/MacOS/Google Chrome Canary");
+	if (args.headless) {
+	options = options.addArguments('--headless');
```
Comment (agubler) [inline tests/benchmark/runner/src/benchmarkRunner.ts]: formatting

File: tests/benchmark/runner/src/benchmarkRunner.ts [hunk 16]
```diff
@@ -0,0 +1,448 @@
...
+	return new Builder()
+		.forBrowser('chrome')
+		.setChromeOptions(options)
+		.build();
+}
+
+async function forceGC(framework: FrameworkData, driver: WebDriver): Promise<any> {
+	if (startsWith(framework.name, 'angular-v4')) {
+		// workaround for window.gc for angular 4 - closure rewrites windows.gc");
+		await driver.executeScript('window.Angular4PreservedGC();');
```
Comment (agubler) [inline tests/benchmark/runner/src/benchmarkRunner.ts]: don't think we need this

File: tests/benchmark/runner/src/common.ts [hunk 17]
```diff
@@ -0,0 +1,38 @@
...
+	useShadowRoot: boolean | undefined;
+}
+
+interface Options {
+	uri: string | null;
+	useShadowRoot?: boolean;
+}
+
+function f(name: string, keyed: boolean, options: Options = {uri: null, useShadowRoot: false}): FrameworkData {
+	let ret = {name, keyed, uri: options.uri ? options.uri : name, useShadowRoot: options.useShadowRoot};
```
Comment (agubler) [inline tests/benchmark/runner/src/common.ts]: can we just return this?

File: tests/benchmark/runner/src/createIndex.ts [hunk 18]
```diff
@@ -0,0 +1,12 @@
+import * as fs from 'fs';
```
Comment (agubler) [inline tests/benchmark/runner/src/createIndex.ts]: do we use this?

File: tests/benchmark/runner/src/createResultTable.ts [hunk 19]
```diff
@@ -0,0 +1,204 @@
+import * as _ from 'lodash';
```
Comment (agubler) [inline tests/benchmark/runner/src/createResultTable.ts]: can we replace lodash?

Comment (agubler) [reply tests/benchmark/runner/src/createResultTable.ts]: with modules from shim perhaps?

File: tests/benchmark/runner/src/createResultTable.ts [hunk 20]
```diff
@@ -0,0 +1,204 @@
...
+	let benches: Array<BenchResultList> = [];
+
+	let filteredFrameworks = frameworks.filter(f => frameworkPredicate(f)).slice();
+
+	let sortedFrameworks = filteredFrameworks.sort((a: FrameworkData, b: FrameworkData) => {
+		if (a.name === referenceName) {
+			if (b.name === referenceName) { return 0; }
+			console.log('found reference name', referenceName);
+			return 1;
+		} else if (b.name === referenceName) {
```
Comment (agubler) [inline tests/benchmark/runner/src/createResultTable.ts]: a hangover from the original code, should we apply the dojo 2 style guide rules? like else on a new line

File: tests/benchmark/runner/src/nonKeyed.ts [hunk 21]
```diff
@@ -0,0 +1,179 @@
+import * as chrome from 'selenium-webdriver/chrome';
```
Comment (agubler) [inline tests/benchmark/runner/src/nonKeyed.ts]: don't think we need this (just checks that the implementation is keyed or not I think)

File: tests/benchmark/runner/src/runCheck.ts [hunk 22]
```diff
@@ -0,0 +1,77 @@
+import * as fs from 'fs';
```
Comment (agubler) [inline tests/benchmark/runner/src/runCheck.ts]: does this get used?

File: tests/benchmark/css/github-markdown.css (hunk 11 above)
Comment (umaar) [reply tests/benchmark/css/github-markdown.css]: Removed

File: tests/benchmark/runner/src/benchmarkRunner.ts (hunk 12 above)
Comment (umaar) [reply tests/benchmark/runner/src/benchmarkRunner.ts]: Done

File: tests/benchmark/runner/src/benchmarkRunner.ts (hunk 13 above)
Comment (umaar) [reply tests/benchmark/runner/src/benchmarkRunner.ts]: Done

File: tests/benchmark/runner/src/benchmarkRunner.ts (hunk 14 above)
Comment (umaar) [reply tests/benchmark/runner/src/benchmarkRunner.ts]: Done

File: tests/benchmark/runner/src/benchmarkRunner.ts (hunk 15 above)
Comment (umaar) [reply tests/benchmark/runner/src/benchmarkRunner.ts]: Done

File: tests/benchmark/runner/src/benchmarkRunner.ts (hunk 16 above)
Comment (umaar) [reply tests/benchmark/runner/src/benchmarkRunner.ts]: Removed

File: tests/benchmark/runner/src/common.ts (hunk 17 above)
Comment (umaar) [reply tests/benchmark/runner/src/common.ts]: Done

File: tests/benchmark/runner/src/createIndex.ts (hunk 18 above)
Comment (umaar) [reply tests/benchmark/runner/src/createIndex.ts]: Removed

File: tests/benchmark/runner/src/createResultTable.ts (hunk 19 above)
Comment (umaar) [reply tests/benchmark/runner/src/createResultTable.ts]: Done

File: tests/benchmark/runner/src/createResultTable.ts (hunk 20 above)
Comment (umaar) [reply tests/benchmark/runner/src/createResultTable.ts]: I would recommend we don't use code review for applying style guide rules, specifically those which should be automated (so builds should fail is they do not match style guide rules)

File: tests/benchmark/runner/src/nonKeyed.ts (hunk 21 above)
Comment (umaar) [reply tests/benchmark/runner/src/nonKeyed.ts]: Removed

File: tests/benchmark/runner/src/runCheck.ts (hunk 22 above)
Comment (umaar) [reply tests/benchmark/runner/src/runCheck.ts]: Looks like it doesn't, removed

**Comment by umaar** [comment]
@agubler have done a rebase with master to make sure the branch is up to date. Another sanity check would be useful 👍

I would also suggest we automate as many style guide related issues as possible. So things should fail if (for example) there are one too many new lines. IMO it doesn't really make sense for humans to be looking out for these things (this can also help with less cognitive load for the reviewer & original developer), rather robots should be checking these!

**Comment by agubler** [comment]
@umaar Thanks. I agree re the style guide issues, I think that there are just some things that cannot currently be detected by tslint... If we move to prettier for the formatting rules it will help (and we'll need to update our style guide in line with the rules from prettier)

**Comment by umaar** [comment]
@agubler maybe we could extend from others? https://github.com/buzinas/tslint-eslint-rules/blob/master/src/docs/rules/braceStyleRule.md

In the mean time I've updated else to be on a new line

**Comment by agubler** [comment]
@umaar I think there are various other open issues regarding the code styling and tslint configuration where that discussion should be had :)

**Pull request merged**